File: iOS/Crisis Connect/Features/Messaging/AuthorityMlsCallGate.swift
import Foundation

/// Requires the exact already-approved AuthorityChat device sets before an authority call may ring.
enum AuthorityMlsCallGate {
    static func isVerified(
        selfUid: String,
        peerUid: String,
        scopeType: AuthorityMlsScopeType,
        channelId: String
    ) async -> Bool {
        do {
            let canonical = try AuthorityMlsIdentifiers.canonicalBinding(AuthorityMlsBinding(
                scopeType: scopeType,
                channelId: channelId,
                participants: [selfUid, peerUid]
            ))
            let conversationId = try AuthorityMlsIdentifiers.conversationId(canonical)
            let directory = try await AuthorityMlsTransport().loadDeviceDirectory(conversationId: conversationId)
            guard directory.rejected == 0 else { return false }
            let grouped = Dictionary(grouping: directory.records, by: \.uid)
            guard Set(grouped.keys) == Set(canonical.participants) else { return false }
            let trust = AuthorityMlsTrustStore()
            return try canonical.participants.allSatisfy { uid in
                try trust.verifyExisting(
                    conversationId: conversationId,
                    uid: uid,
                    devices: grouped[uid] ?? []
                ).approved
            }
        } catch {
            return false
        }
    }
}

File: Android/app/src/main/java/com/auralis/crisisconnect/messaging/AuthorityMlsCallGate.kt
package com.auralis.crisisconnect.messaging

import android.content.Context

/** Requires the exact already-approved AuthorityChat device sets before an authority call may ring. */
object AuthorityMlsCallGate {
    suspend fun isVerified(
        context: Context,
        selfUid: String,
        peerUid: String,
        scopeType: AuthorityMlsScopeType,
        channelId: String,
    ): Boolean = runCatching {
        val canonical = AuthorityMlsIdentifiers.canonicalBinding(
            AuthorityMlsBinding(scopeType, channelId, listOf(selfUid, peerUid)),
        )
        val conversationId = AuthorityMlsIdentifiers.conversationId(canonical)
        val directory = AuthorityMlsTransport().loadDeviceDirectory(conversationId)
        if (directory.rejected != 0) return@runCatching false
        val grouped = directory.records.groupBy { it.uid }
        if (grouped.keys != canonical.participants.toSet()) return@runCatching false
        val trust = AuthorityMlsTrustStore(context.applicationContext)
        canonical.participants.all { uid ->
            trust.verifyExisting(conversationId, uid, grouped[uid].orEmpty()).approved
        }
    }.getOrDefault(false)
}
